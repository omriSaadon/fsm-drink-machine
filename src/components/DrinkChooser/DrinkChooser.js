import React from 'react';
import { Button, Card } from 'antd';
import axios from 'axios';
import { useStore } from 'store/FsmStore';
import { getPaid, getMachineStatus } from 'store/fsm_reducer';
import {
  GENERAL_ACTIONS, STATES, ACTION_STATES, DRINKS,
} from 'constants/index';
import classes from './DrinkChooser.scss';

const DrinkChooser = () => {
  const [state, dispatch] = useStore();

  const paid = getPaid(state);
  const drinkMachineStatus = getMachineStatus(state);

  const onBuy = (price) => () => {
    if (drinkMachineStatus === STATES.DELIVERY) {
      // In this case, purchasing while already delivering other product is forbidden and an error message will appear.
      dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: 'Please wait until delivery is finished', type: 'error' } });
    } else if (paid - price >= 0) {
      // In case we purchased a product but still has paid money to the machine, the state should be Ordering as we can order more products.
      dispatch({ state: ACTION_STATES.PURCHASE, payload: price });
      // Calling to a mock server which has delay of few seconds.
      axios.post('https://fsmmock.free.beeceptor.com', { price })
        .then((res) => {
          dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: res.data, type: 'success' } });
          if (paid - price === 0) dispatch({ state: ACTION_STATES.DONE }); // If paid is zero, we should move to Idle state.
          else dispatch({ state: ACTION_STATES.PAYMENT, payload: 0 }); // If paid is bigger than zero, payment state should take place.
        }).catch(() => {
          dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: 'Machine is out of service, your money is gone', type: 'error' } });
        });
    } else {
      // If trying to purchase a product which we didn't pay enough money for, an error message will appear.
      dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: 'Insufficient funds', type: 'error' } });
    }
  };

  return (
    <Card title="Choose Your Drink">
      <div className={classes.wrapper}>
        <div className={classes.drinks}>
          {DRINKS.map((drink) => (
            <Button key={drink.key} onClick={onBuy(drink.price)} shape="round" size="large" icon={drink.icon}>{`${drink.label} - ${drink.price} â‚ª`}</Button>))}
        </div>
      </div>
    </Card>
  );
};

export default DrinkChooser;
