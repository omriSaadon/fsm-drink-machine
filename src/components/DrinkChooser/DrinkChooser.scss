.drinks {
  width: 250px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  height: 200px;
  justify-content: space-between;
}

.error {
  padding-top: 10px;
}

.wrapper {
  height: 200px;
}