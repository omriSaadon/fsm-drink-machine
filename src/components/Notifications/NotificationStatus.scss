.notification {
  width: calc(100% - 10px);
  margin-bottom: 20px;
  height: 40px;
}