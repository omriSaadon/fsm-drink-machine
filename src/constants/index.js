import React from 'react';

export const INITIAL_BUDGET = 10;

export const GENERAL_ACTIONS = {
  ADD_BUDGET: 'addBudget',
  ADD_MESSAGE: 'addMessage',
};

export const STATES = {
  IDLE: 'idle',
  ORDERING: 'ordering',
  DELIVERY: 'delivery',
};

export const ACTION_STATES = {
  PAYMENT: 'payment',
  CANCEL: 'cancel',
  PURCHASE: 'purchase',
  DONE: 'done',
};

export const COINS = [
  { key: 'one', label: '1 ₪', value: 1 },
  { key: 'two', label: '2 ₪', value: 2 },
  { key: 'five', label: '5 ₪', value: 5 },
  { key: 'ten', label: '10 ₪', value: 10 },
];

export const DRINKS = [
  {
    key: 'Coke', label: 'Coke', price: 8, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/coca-cola.jpeg`} />,
  },
  {
    key: 'Sparkling_Water', label: 'Sparkling Water', price: 6, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/sparkling-water.jpeg`} />,
  },
  {
    key: 'Water', label: 'Water', price: 5, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/water.jpeg`} />,
  },
  {
    key: 'Ice_Tea', label: 'Ice Tea', price: 7, icon: <img alt="" src={`${process.env.PUBLIC_URL}/assets/images/ice-tea.jpeg`} />,
  },
];

export const FLOW_INITIAL_ELEMENTS = [
  {
    id: STATES.IDLE,
    data: { label: 'Idle' },
    position: { x: 10, y: 100 },
    targetPosition: 'right',
    sourcePosition: 'bottom',
  },
  {
    id: 'payment_needed', data: { label: 'Payment Is Needed' }, position: { x: 300, y: 10 }, targetPosition: 'left', sourcePosition: 'left',
  },
  {
    id: STATES.ORDERING, data: { label: 'Ordering' }, position: { x: 420, y: 240 }, sourcePosition: 'left',
  },
  {
    id: STATES.DELIVERY, data: { label: 'delivery' }, position: { x: 550, y: 100 }, sourcePosition: 'left', targetPosition: 'bottom',
  },
  {
    id: ACTION_STATES.CANCEL, data: { label: 'Cancel Order' }, position: { x: 100, y: 200 }, targetPosition: 'right', sourcePosition: 'top',
  },
  {
    id: 'idle-payment_needed',
    source: STATES.IDLE,
    target: 'payment_needed',
    labelStyle: { fill: 'red', fontWeight: 700 },
    label: 'Clicked item without payment',
  },
  {
    id: 'idle-ordering', source: STATES.IDLE, target: STATES.ORDERING, animated: true, label: 'Payment Succeed', arrowHeadType: 'arrowclosed',
  },
  {
    id: 'ordering-delivery', source: STATES.ORDERING, target: STATES.DELIVERY, animated: true, label: 'Item Selected', arrowHeadType: 'arrowclosed',
  },
  {
    id: 'ordering-cancel',
    source: STATES.ORDERING,
    target: ACTION_STATES.CANCEL,
    label: 'Canceled',
    labelStyle: { fill: 'red', fontWeight: 700 },
    arrowHeadType: 'arrowclosed',
  },
  {
    id: 'cancel-idle', source: ACTION_STATES.CANCEL, target: STATES.IDLE, arrowHeadType: 'arrowclosed',
  },
  {
    id: 'delivery-idle',
    source: STATES.DELIVERY,
    target: STATES.IDLE,
    arrowHeadType: 'arrowclosed',
    animated: true,
    labelBgPadding: [8, 4],
    labelBgBorderRadius: 4,
    labelBgStyle: {
      fill: '#FFCC00', fillOpacity: 1,
    },
    label: 'Finished',
  },
];
