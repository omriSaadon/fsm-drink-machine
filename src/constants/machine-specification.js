import {
  makePayment, addBudget, cancelPayment, makePurchase, addMessage,
} from 'store/fsm_reducer';
import { INITIAL_BUDGET, STATES, ACTION_STATES } from '.';

export default {
  initialState: {
    state: STATES.IDLE,
    context: {
      currentBudget: INITIAL_BUDGET,
      paid: 0,
      message: '',
    },
  },
  generalActions: {
    addBudget,
    addMessage,
  },
  states: {
    [STATES.IDLE]: {
      [ACTION_STATES.PAYMENT]: {
        next: STATES.ORDERING,
        action: makePayment,
      },
    },
    [STATES.ORDERING]: {
      [ACTION_STATES.CANCEL]: {
        next: STATES.IDLE,
        action: cancelPayment,
      },
      [ACTION_STATES.PURCHASE]: {
        next: STATES.DELIVERY,
        action: makePurchase,
      },
      [ACTION_STATES.PAYMENT]: {
        next: STATES.ORDERING,
        action: makePayment,
      },
    },
    [STATES.DELIVERY]: {
      [ACTION_STATES.DONE]: {
        next: STATES.IDLE,
      },
      [ACTION_STATES.PAYMENT]: {
        next: STATES.ORDERING,
        action: makePayment,
      },
    },
  },
};
