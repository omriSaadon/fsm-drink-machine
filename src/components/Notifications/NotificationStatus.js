import React, { useEffect } from 'react';
import { Alert } from 'antd';
import { useStore } from 'store/FsmStore';
import { getMessageType, getMessage } from 'store/fsm_reducer';
import { GENERAL_ACTIONS } from 'constants/index';
import classes from './NotificationStatus.scss';

const NotificationStatus = () => {
  const [state, dispatch] = useStore();

  const message = getMessage(state);
  const messageType = getMessageType(state);

  const onClose = () => {
    dispatch({ generalAction: GENERAL_ACTIONS.ADD_MESSAGE, payload: { message: '', type: null } });
  };

  useEffect(() => {
    const timer = setTimeout(() => {
      onClose();
    }, 5000);
    return () => clearTimeout(timer);
  }, [message]);

  return (
    <div className={classes.notification}>
      {message
            && (
            <Alert
              message={message}
              type={messageType}
              closable
              onClose={onClose}
            />
            )}
    </div>

  );
};

export default NotificationStatus;
