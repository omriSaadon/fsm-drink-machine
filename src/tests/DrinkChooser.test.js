import React from 'react';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import axios from 'axios';
import DrinkChooser from 'components/DrinkChooser/DrinkChooser';
import NotificationStatus from 'components/Notifications/NotificationStatus';
import { STATES } from 'constants/index';
import { machineRender } from './helpers/machineHelper';
import mockStore from './mocks/mock_context';

jest.mock('axios', () => ({
  post: jest.fn(() => new Promise((resolve) => {
    resolve(true);
  })),
}));

test('Check Insufficient funds', () => {
  machineRender(
    <>
      <NotificationStatus />
      <DrinkChooser />
    </>,
  );
  const button = screen.getByRole('button', { name: 'Coke - 8 ₪' });
  fireEvent.click(button);
  const linkElement = screen.queryByText(/Insufficient funds/i);
  expect(linkElement).toBeInTheDocument();
});

test('Check when chose drink while delivering is in progress', () => {
  machineRender(
    <>
      <NotificationStatus />
      <DrinkChooser />
    </>,
    { spec: mockStore({ state: STATES.DELIVERY }) },
  );
  const button = screen.getByRole('button', { name: 'Coke - 8 ₪' });
  fireEvent.click(button);
  const linkElement = screen.queryByText(/Please wait until delivery is finished/i);
  expect(linkElement).toBeInTheDocument();

  const notification = screen.getByRole('button', { name: 'close' });
  fireEvent.click(notification);

  const errorElement = screen.queryByText(/Please wait until delivery is finished/i);
  expect(errorElement).not.toBeInTheDocument();
});

test('Check successful delivery and staying in payment', async () => {
  axios.post.mockResolvedValue({ data: 'Enjoy !' });
  machineRender(
    <>
      <NotificationStatus />
      <DrinkChooser />
    </>,
    { spec: mockStore({ state: STATES.ORDERING, paid: 20 }) },
  );

  const button = screen.getByRole('button', { name: 'Coke - 8 ₪' });
  fireEvent.click(button);
  await waitFor(() => {
    const linkElement = screen.queryByText(/Enjoy !/i);
    expect(linkElement).toBeInTheDocument();
    expect(axios.post).toHaveBeenCalledTimes(1);
  });
});

test('Check successful delivery and moving to done', async () => {
  axios.post.mockResolvedValue({ data: 'Enjoy !' });
  machineRender(
    <>
      <NotificationStatus />
      <DrinkChooser />
    </>,
    { spec: mockStore({ state: STATES.ORDERING, paid: 8 }) },
  );

  const button = screen.getByRole('button', { name: 'Coke - 8 ₪' });
  fireEvent.click(button);
  await waitFor(() => {
    const linkElement = screen.queryByText(/Enjoy !/i);
    expect(linkElement).toBeInTheDocument();
    expect(axios.post).toHaveBeenCalledTimes(1);
  });
});

test('Check server failure', async () => {
  axios.post.mockRejectedValueOnce();
  machineRender(
    <>
      <NotificationStatus />
      <DrinkChooser />
    </>,
    { spec: mockStore({ state: STATES.ORDERING, paid: 8 }) },
  );

  const button = screen.getByRole('button', { name: 'Coke - 8 ₪' });
  fireEvent.click(button);
  await waitFor(() => {
    const linkElement = screen.queryByText(/Machine is out of service, your money is gone/i);
    expect(linkElement).toBeInTheDocument();
  });
});

test('Verify snapshot', () => {
  const { asFragment } = machineRender(<DrinkChooser />);
  expect(asFragment()).toMatchSnapshot();
});
