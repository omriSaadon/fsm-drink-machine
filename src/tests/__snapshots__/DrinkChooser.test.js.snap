// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          Choose Your Drink
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div>
        <div>
          <button
            class="ant-btn ant-btn-round ant-btn-lg"
            type="button"
          >
            <img
              alt=""
              src="undefined/assets/images/coca-cola.jpeg"
            />
            <span>
              Coke - 8 ₪
            </span>
          </button>
          <button
            class="ant-btn ant-btn-round ant-btn-lg"
            type="button"
          >
            <img
              alt=""
              src="undefined/assets/images/sparkling-water.jpeg"
            />
            <span>
              Sparkling Water - 6 ₪
            </span>
          </button>
          <button
            class="ant-btn ant-btn-round ant-btn-lg"
            type="button"
          >
            <img
              alt=""
              src="undefined/assets/images/water.jpeg"
            />
            <span>
              Water - 5 ₪
            </span>
          </button>
          <button
            class="ant-btn ant-btn-round ant-btn-lg"
            type="button"
          >
            <img
              alt=""
              src="undefined/assets/images/ice-tea.jpeg"
            />
            <span>
              Ice Tea - 7 ₪
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
